cos-fleetshard-sync/src/main/java/org/bf2/cos/fleetshard/sync/housekeeping/reapers/AddonReaper.java

```diff
@@ -0,0 +1,183 @@
+                "All namespaces have been deleted. Deleting {} in namespace {} to signal that addon " +
+                    "and operators removals should proceed.",
+                config.addon().olmOperatorsKind(),
+                config.namespace());
+
+            deleteAddonResource();
+            LOGGER.info("Cluster cleanup should be completed. Addon and Operators removals should proceed normally.");
+        } else {
+            LOGGER.info("Namespaces still not deleted, will wait for their deletion and try again.");
+        }
+    }
+
+    private void deleteAddonResource() {
+        ResourceDefinitionContext rdc = new ResourceDefinitionContext.Builder()
+            .withGroup(config.addon().olmOperatorsGroup())
+            .withVersion(config.addon().olmOperatorsApiVersion())
+            .withKind(config.addon().olmOperatorsKind())
+            .build();
+
+        kubernetesClient.genericKubernetesResources(rdc)
+            .inNamespace(config.namespace())
+            .delete();
+    }
+
+    private FilterWatchListDeletable<Namespace, NamespaceList> getNamespaceFilter() {
+        return kubernetesClient.namespaces().withLabel(Resources.LABEL_CLUSTER_ID, config.cluster().id());
+    }
+
+    public class ConfigMapWatcher extends AbstractWatcher<ConfigMap> {
+        private static final String LABEL_PREFIX = "api.openshift.com/";
+        private static final String DELETE_LABEL_SUFFIX = "-delete";
+
+        @Override
+        protected Watch doWatch() {
+            LOGGER.info("Creating Watcher for Addon ConfigMaps on namespace {} with Addon ID {}",
+                config.namespace(), config.addon().id());
+
+            return kubernetesClient
+                .configMaps()
+                .inNamespace(config.namespace())
+                .withName(config.addon().id())
+                .watch(this);
+        }
+
+        @Override
+        protected void onEventReceived(Action action, ConfigMap resource) {
+            LOGGER.info("Event received for Addon ConfigMap with action {}", action.name());
+            switch (action) {
+                case ADDED:
+                case MODIFIED:
+                    cleanupCluster(resource);
+                    break;
+                case DELETED:
+                case ERROR:
+                case BOOKMARK:
+                    LOGGER.debug("Event {} not covered by this watcher, nothing will be done.", action.name());
+                    break;
+            }
+        }
+
+        private void cleanupCluster(ConfigMap configMap) {
+            final String labelName = LABEL_PREFIX + config.addon().label() + DELETE_LABEL_SUFFIX;
+            final String deleteLabel = Resources.getLabel(configMap, labelName);
+
+            if ("true".equals(deleteLabel)) {
+                LOGGER.info("ConfigMap for deletion of the Addon was found, starting cleanup of cluster: {} - {}",
+                    config.cluster().id(), configMap);
+                fleetShardSync.stop();
```

Comment:
If we stop everything here, including the `HouseKeeper`, who is going to keep calling `run()` so we have a retry loop?

Reply:
good catch